File: src/cpp/sliding_window/3.longest-substring-without-repeating-characters.cpp
/*
 * @lc app=leetcode id=3 lang=cpp
 *
 * [3] Longest Substring Without Repeating Characters
 *
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
 */

#include <algorithm>
#include <string>
#include <unordered_map>
using namespace std;

// @lc code=start
class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    size_t i = 0;
    int l = 0;
    int maxlen = 0;
    unordered_map<char, int> charind;
    while (i < s.size()) {
      char ch = s[i];
      if (charind.count(ch) && charind[ch] >= l) {
        l = charind[ch] + 1;
      }
      charind[ch] = i;
      int wind = i - l + 1;
      maxlen = std::max(maxlen, wind);
      i++;
    }
    return maxlen;
  }
};
// @lc code=end

// Non-LeetCode wrapper for harness
int lengthOfLongestSubstring(string s) {
  Solution sol;
  return sol.lengthOfLongestSubstring(std::move(s));
}

File: src/java/sliding_window/3.longest-substring-without-repeating-characters.java
class P3 {
    // @lc app=leetcode id=3 lang=java
    // @lc code=start
    public class Solution {
        public int lengthOfLongestSubstring(String s) {
            int[] last = new int[256];
            java.util.Arrays.fill(last, -1);
            int best = 0, start = 0;
            for (int i = 0; i < s.length(); i++) {
                int c = s.charAt(i);
                if (last[c] >= start) start = last[c] + 1;
                last[c] = i;
                best = Math.max(best, i - start + 1);
            }
            return best;
        }
    }
    // @lc code=end
}
